#  Work on Data preparation, visualization and Machine learning with Python
## Data Science in Production

<img src='https://www.uniquindio.edu.co/info/uniquindio/media/bloque2.png' width="90" height="110" >

**Students:** <br>

Humberto Franco Osorio - Juan Felipe Padilla


**Email:** hfranoo@uqvirtual.edu.co - jfpadillag@uqvirtual.edu.co


**Teacher:** [Jose R. Zapata](https://joserzapata.github.io) <br>
<br>
<a href='https://joserzapata.github.io'>
    <img src='https://1000marcas.net/wp-content/uploads/2020/02/logo-GitHub.png' width="90" height="50" >
</a>
&nbsp;
<a href='https://twitter.com/joserzapata'>
    <img src='https://upload.wikimedia.org/wikipedia/commons/thumb/4/4f/Twitter-logo.svg/1200px-Twitter-logo.svg.png' width="50" height="40" >
</a>
&nbsp;&nbsp;&nbsp;&nbsp;
<a href='https://www.linkedin.com/in/jose-ricardo-zapata-gonzalez/'>
    <img src='https://cdn-icons-png.flaticon.com/512/174/174857.png' width="50" height="50" >
</a>


## Table of Contents
1) Import Libreries
2) Load Dataset
3) Data description and cleaning

    - Variable identification
    - Missing  and  Duplicate values
        - Identify missing values
        - Identify duplicated
        - Dataset cleaning
    - Elimination of features that do not provide information
4) Save intermediated transformed data
5) Partial Results

### Introduction
Analyze the relationship between the characteristics of a cell phone and its selling price, with the aim of classifying the price range in which a cell phone is found.

4 different outputs: <br>
&nbsp;&nbsp;&nbsp; **0** -> Low  <br>
&nbsp;&nbsp;&nbsp; **1** -> Medium   <br>
&nbsp;&nbsp;&nbsp; **2** -> High   <br>
&nbsp;&nbsp;&nbsp; **3** -> Very High   <br>

### Main Objective
> What is the price range of a mobile phone according to its features?

## 1. Import Libreries

In [22]:
import pandas as pd
pd.options.plotting.backend = "plotly"
import numpy as np
import re

## 2. Load Dataset

In [23]:
ds = pd.read_csv("../data/raw/CellPhoneDS.csv")

## 3. Data description and cleaning

### Variable identification

_**Table 1.** List of features and their descriptions in the initial dataset._

|     | Feature name  | Description and values |  Type| % missing |
|:---:|     :---:     | :---: |:---: |:---: | 
| 1 | Unnamed: 0    | Unique identifier of registered mobile phone. | Numeric |  0
| 2 | index         | Unique mobile phone model identifier. | Nominal | 0
| 3 | battery_power | Total energy a battery can store in one time measured in mAh.     | Numeric | 3.77
| 4 | blue          | Has bluetooth or not. <br>0: NO. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  1: YES.| Nominal| 4.02
| 5 | clock_speed   | Speed at which microprocessor executes instructions.<br> 31 distinct values. | Numeric| 3.67
| 6 | dual_sim      | Has dual sim support or not.<br>0: NO. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  1: YES. | Nominal | 3.54
| 7 | fc            | Front Camera mega pixels. | Numeric| 4.00
| 8 | four_g        | Has 4G or not. <br>0: NO. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  1: YES.| Nominal | 3.96
| 9 | int_memory    | Internal Memory in Gigabytes. | Numeric | 3.86
| 10 | m_dep         | Mobile Depth in cm. | Numeric  | 3.79
| 11 | mobile_wt     | Weigh of mobile phone. | Numeric | 3.67
| 12 | n_cores       | Number of cores of processor. | Numeric| 3.71
| 13 | pc            | Primary Camera mega pixels. | Numeric| 3.54
| 14 | px_height     | Pixel Resolution Height. | Numeric | 3.86
| 15 | px_width      | Pixel Resolution Width | Numeric | 3.77
| 16 | ram           | Random Access Memory in Megabytes.| Numeric  | 4.02
| 17 | sc_h          | Screen Height of mobile in cm.| Numeric  | 3.87
| 18 | sc_w          | Screen Width of mobile in cm. | Numeric  | 3.58
| 19 | talk_time     | Longest time that a single battery charge will last when you are charget.  | Numeric | 4.01
| 20 | three_g       | Has 3G or not. <br>0: NO. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  1: YES.| Nominal | 3.64
| 21 | touch_screen  | Has touch screen or not. <br>0: NO. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  1: YES.| Nominal | 3.56
| 22 | wifi          | Has wifi or not. <br>0: NO. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  1: YES.| Nominal | 3.64
| 23 | hgv           | <font color='yellow'>Feature with no information, no information can be collected for this feature.</font> | Numeric | 3.74
| 24 | werf          | <font color='yellow'>Feature with no information, no information can be collected for this feature.</font> | Nominal | 0
| 25 | price_range   | Price range is given by categories.<br> &nbsp;&nbsp;&nbsp; **0** -> Low  <br>&nbsp;&nbsp;&nbsp; **1** -> medium   <br>&nbsp;&nbsp;&nbsp; **2** -> High   <br>&nbsp;&nbsp;&nbsp; **3** -> Very High   <br> | Ordinal | 4.58

The initial working dataset has 25 columns (feature) and 74790 registered mobile phones.

In [24]:
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74790 entries, 0 to 74789
Data columns (total 25 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     74790 non-null  int64 
 1   index          74790 non-null  int64 
 2   talk_time      71990 non-null  object
 3   battery_power  71970 non-null  object
 4   pc             72140 non-null  object
 5   three_g        72070 non-null  object
 6   mobile_wt      72050 non-null  object
 7   px_width       71970 non-null  object
 8   sc_h           71895 non-null  object
 9   sc_w           72115 non-null  object
 10  m_dep          71955 non-null  object
 11  touch_screen   72130 non-null  object
 12  fc             71800 non-null  object
 13  four_g         71825 non-null  object
 14  hgv            71990 non-null  object
 15  price_range    71365 non-null  object
 16  blue           71785 non-null  object
 17  n_cores        72015 non-null  object
 18  wifi           72070 non-n

### Missing  and  Duplicate values

#### A) Identify missing values

In [25]:
ds.sample(10).T

68724    20242            47583    22573  \
Unnamed: 0               68724    20242            47583    22573   
index                      777     1801             2455      212   
talk_time                  7.0      NaN             12.0     13.0   
battery_power           1786.0   1035.0            537.0   1049.0   
pc                        10.0     10.0              2.0     17.0   
three_g                    1.0      1.0  -948961565145.0      NaN   
mobile_wt                161.0    173.0            103.0    135.0   
px_width                1671.0   1709.0           1430.0   1077.0   
sc_h                       9.0      NaN             10.0     11.0   
sc_w                       3.0      8.0              5.0      3.0   
m_dep                      0.7      1.0              0.3      NaN   
touch_screen               1.0      1.0              1.0      0.0   
fc                         9.0      NaN              1.0      9.0   
four_g         -948961565145.0      0.0              1.0      0.0   
hgv                        7.0      NaN             12.0     13.0   
price_range                3.0      NaN  -948961565145.0      0.0   
blue                       0.0      1.0              1.0      NaN   
n_cores                    7.0      NaN              7.0      5.0   
wifi                       1.0      NaN              1.0      0.0   
dual_sim                   0.0      NaN              0.0      0.0   
ram                     2764.0   2002.0  -948961565145.0   1022.0   
werf                   9156615  9156615          9156615  9156615   
int_memory                58.0      NaN             55.0     63.0   
px_height                231.0    831.0  -948961565145.0    199.0   
clock_speed                0.5      NaN              2.0      2.2   

                          23452    56700             3546             51781  \
Unnamed: 0                23452    56700              3546            51781   
index                      1086     2034               750             1634   
talk_time                  19.0     15.0               8.0             18.0   
battery_power  nhbgvfrtd 56gyub    543.0            1276.0            986.0   
pc                         13.0      4.0               3.0             20.0   
three_g        nhbgvfrtd 56gyub      1.0               1.0              1.0   
mobile_wt                 152.0    192.0             130.0            182.0   
px_width       nhbgvfrtd 56gyub      NaN  nhbgvfrtd 56gyub            679.0   
sc_h                       15.0      NaN              19.0              8.0   
sc_w                        3.0     14.0              10.0              7.0   
m_dep                       0.4      0.7  nhbgvfrtd 56gyub              0.8   
touch_screen   nhbgvfrtd 56gyub      0.0               1.0              1.0   
fc                          0.0      0.0               2.0             13.0   
four_g                      1.0      0.0               0.0              1.0   
hgv                        19.0     15.0               8.0             18.0   
price_range                 1.0      1.0               3.0              0.0   
blue                        0.0      NaN  nhbgvfrtd 56gyub              1.0   
n_cores                     1.0      5.0               2.0              4.0   
wifi                        0.0      0.0               1.0              0.0   
dual_sim       nhbgvfrtd 56gyub      0.0               1.0              0.0   
ram                      1375.0      NaN            3798.0           1149.0   
werf                    9156615  9156615           9156615          9156615   
int_memory                 30.0     57.0              53.0             28.0   
px_height                 163.0    391.0  nhbgvfrtd 56gyub            199.0   
clock_speed                 2.6      0.5               2.9  5285988458456.0   

                         74567            55246  
Unnamed: 0               74567            55246  
index                      854             2492  
talk_time                

The presence of values such as '??????', 'nhbgvfrtd 56gyub', or '-948961565145.0' is observed, as well as values for some features that are totally atypical considering the description of the feature, is important to find which values had every column.

Get unique values per columns.

In [26]:
print("-----------------------------------------------------------------------------------")
for col in ds:
    print(f"{col} = {ds[col].unique()}")
    print("Number option",f"{len(ds[col].unique())}")
    print("-----------------------------------------------------------------------------------")

-----------------------------------------------------------------------------------
Unnamed: 0 = [    0     1     2 ... 74787 74788 74789]
Number option 74790
-----------------------------------------------------------------------------------
index = [1517  911 1493 ... 1192 1495 1181]
Number option 2493
-----------------------------------------------------------------------------------
talk_time = ['11.0' '9.0' '3.0' '-948961565145.0' '6.0' 'nhbgvfrtd 56gyub' '12.0'
 '8.0' '16.0' '20.0' '13.0' '5285988458456.0' '14.0' '7.0' nan '19.0'
 '10.0' '15.0' '4.0' '18.0' '??????' '17.0' '5.0' '2.0']
Number option 24
-----------------------------------------------------------------------------------
battery_power = ['911.0' '1284.0' '1183.0' ... '527.0' '1542.0' '846.0']
Number option 1099
-----------------------------------------------------------------------------------
pc = ['4.0' '14.0' 'nhbgvfrtd 56gyub' '9.0' '11.0' '20.0' '1.0' '??????' '0.0'
 '16.0' '7.0' '5.0' '5285988458456.0' '10.0' 

**Analysis**

- Most of the features have invalid values such as 'nhbgvfrtd 56gyub', '??????' and 'nan'.
- Most of the characteristics have invalid values considering the description of these values, such as: '-948961565145.0' y '5285988458456.0'
- The column 'werf' has the same value in all rows (9156615). So do not provide any information for the classifier.


In [27]:
ds.drop('werf', axis=1, inplace=True)

From a previous analysis of the data set, we observed the presence of outliers registered as:
- 'nhbgvfrtd 56gyub'
- '??????'
- 'nan'
- '-948961565145.0'
- '5285988458456.0'

In [28]:
missing_value = ['nhbgvfrtd 56gyub', '??????', 'nan', '-948961565145.0', '5285988458456.0']

Analyzing the presence of these values in the different characteristics of the dataset.

In [29]:
print("-------------------------------------------------------------")
for i in range(len(missing_value)):
    print("Valores iguales a ",missing_value[i])
    frec=ds[ds == missing_value[i]].count()
    for col in ds:
        if frec[col] > 0:
            print(f"{col} = {frec[col]}")
    print("-------------------------------------------------------------")

-------------------------------------------------------------
Valores iguales a  nhbgvfrtd 56gyub
talk_time = 2495
battery_power = 2495
pc = 2495
three_g = 2495
mobile_wt = 2495
px_width = 2495
sc_h = 2495
sc_w = 2495
m_dep = 2495
touch_screen = 2495
fc = 2495
four_g = 2495
hgv = 2495
price_range = 2495
blue = 2495
n_cores = 2495
wifi = 2495
dual_sim = 2495
ram = 2495
int_memory = 2495
px_height = 2495
clock_speed = 2495
-------------------------------------------------------------
Valores iguales a  ??????
talk_time = 2495
battery_power = 2495
pc = 2495
three_g = 2495
mobile_wt = 2495
px_width = 2495
sc_h = 2495
sc_w = 2495
m_dep = 2495
touch_screen = 2495
fc = 2495
four_g = 2495
hgv = 2495
price_range = 2495
blue = 2495
n_cores = 2495
wifi = 2495
dual_sim = 2495
ram = 2495
int_memory = 2495
px_height = 2495
clock_speed = 2495
-------------------------------------------------------------
Valores iguales a  nan
-------------------------------------------------------------
Valores igual

There is a low presence of these meaningless vouchers in the different characteristics, therefore it is decided to impute these values as NaN values.

In [30]:
for missing in missing_value:
    ds.replace(missing, np.nan, inplace=True)

The percentage of null values in each characteristic increases by approximately 9 points, now there is an average percentage of missing values of 13%, which is a low value considering the size of the dataset, therefore it is decided to eliminate the records with a presence of null values.

In [31]:
(ds.isnull().sum()).plot.bar(title = 'Percentage of Missing Values per column',color=ds.columns)

In [32]:
missing = ds.isnull().sum()
missing[missing>0]*100/len(ds)

talk_time        17.087846
battery_power    17.114588
pc               16.887284
three_g          16.980880
mobile_wt        17.007621
px_width         17.114588
sc_h             17.214868
sc_w             16.920711
m_dep            17.134644
touch_screen     16.900655
fc               17.341891
four_g           17.308464
hgv              17.087846
price_range      17.923519
blue             17.361947
n_cores          17.054419
wifi             16.980880
dual_sim         16.880599
ram              17.361947
int_memory       17.201498
px_height        17.201498
clock_speed      17.014307
dtype: float64

The columns present a percentage of missing values of 17%, therefore none is a candidate for elimination.

#### B) Identify duplicated

Knowing that '*_index_*' is a unique indicator of the mobile phone model, it is assumed that mobile phones with the same model number will have the same characteristics.

With this in mind, we now check whether the model number of the cell phone is registered more than once.

In [33]:
print( 'Duplicated Registers: ' + str( ds["index"].duplicated().sum() ))
print( 'Unique Registers: ' + str( len ( ds["index"].unique() ) ) )

Duplicated Registers: 72297
Unique Registers: 2493


There are 2493 unique values in the '*_index_*', which indicates that the dataset has only 2493 mobile phone models and 72297 duplicate records. Therefore these 72297 duplicate records will be eliminated.

#### C) Dataset cleaning
Taking into account the high number of duplicate records, it is decided to eliminate all the records with NaM values, verifying that the number of unique records (2493) is maintained once these records have been eliminated.

In [34]:
ds.dropna(inplace=True,axis=0)
ds.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12130 entries, 0 to 74782
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     12130 non-null  int64 
 1   index          12130 non-null  int64 
 2   talk_time      12130 non-null  object
 3   battery_power  12130 non-null  object
 4   pc             12130 non-null  object
 5   three_g        12130 non-null  object
 6   mobile_wt      12130 non-null  object
 7   px_width       12130 non-null  object
 8   sc_h           12130 non-null  object
 9   sc_w           12130 non-null  object
 10  m_dep          12130 non-null  object
 11  touch_screen   12130 non-null  object
 12  fc             12130 non-null  object
 13  four_g         12130 non-null  object
 14  hgv            12130 non-null  object
 15  price_range    12130 non-null  object
 16  blue           12130 non-null  object
 17  n_cores        12130 non-null  object
 18  wifi           12130 non-n

By eliminating the records with NaN values we were able to reduce the size of the data set to 12130 records.

In [35]:
ds.sample(10).T

10762   62633   59993   7813    42021   55081   49100   32064  \
Unnamed: 0      10762   62633   59993    7813   42021   55081   49100   32064   
index            1566     729    1537     962    1270    2069     437    1970   
talk_time        11.0     6.0    19.0     7.0    14.0     7.0     7.0     8.0   
battery_power  1317.0  1135.0   733.0  1660.0  1053.0   625.0  1472.0  1913.0   
pc                9.0    11.0    12.0     6.0    17.0     8.0    10.0     4.0   
three_g           1.0     1.0     1.0     0.0     0.0     1.0     1.0     1.0   
mobile_wt       154.0   158.0   138.0   184.0   126.0   183.0   130.0   111.0   
px_width        542.0  1589.0  1787.0  1165.0  1261.0   842.0  1744.0   742.0   
sc_h              7.0    18.0    12.0    12.0    17.0     9.0    12.0    17.0   
sc_w              0.0    13.0     0.0     0.0     0.0     1.0    11.0    13.0   
m_dep             0.1     0.4     0.4     0.8     0.9     0.2     0.5     0.6   
touch_screen      1.0     0.0     1.0     0.0     0.0     1.0     1.0     1.0   
fc                6.0     9.0     7.0     4.0     7.0     5.0     1.0     0.0   
four_g            0.0     0.0     0.0     0.0     0.0     1.0     0.0     0.0   
hgv              11.0     6.0    19.0     7.0    14.0     7.0     7.0     8.0   
price_range       0.0     3.0     2.0     1.0     0.0     2.0     1.0     2.0   
blue              0.0     1.0     0.0     0.0     0.0     1.0     0.0     1.0   
n_cores           2.0     1.0     4.0     1.0     6.0     4.0     6.0     5.0   
wifi              1.0     0.0     0.0     0.0     1.0     1.0     0.0     0.0   
dual_sim          0.0     1.0     0.0     1.0     0.0     1.0     1.0     0.0   
ram             425.0  3204.0  2648.0  1531.0  1339.0  3323.0   985.0  2023.0   
int_memory       12.0    43.0    16.0     7.0    42.0    10.0    36.0    29.0   
px_height       431.0   690.0   202.0    76.0   419.0   635.0  1452.0   675.0   
clock_speed       1.5     2.8     0.8     2.8     2.9     1.2     2.8     1.8   

                53252   60175  
Unnamed: 0      53252   60175  
index            1131    1933  
talk_time        18.0     7.0  
battery_power   831.0   897.0  
pc                9.0    10.0  
three_g           1.0     1.0  
mobile_wt       134.0   154.0  
px_width       1490.0   994.0  
sc_h             16.0     7.0  
sc_w             12.0     5.0  
m_dep             0.7     0.6  
touch_screen      0.0     1.0  
fc                2.0     3.0  
four_g            1.0     1.0  
hgv              18.0     7.0  
price_range       2.0     1.0  
blue              1.0     1.0  
n_cores           8.0     8.0  
wifi              1.0     0.0  
dual_sim          0.0     0.0  
ram            2573.0  1958.0  
int_memory       62.0     2.0  
px_height      1482.0     0.0  
clock_speed       0.7     2.0

Now, we proceed to delete duplicate records in the dataset

In [36]:
ds.drop_duplicates(subset ="index", keep = 'first', inplace = True)
ds.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2328 entries, 0 to 14950
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     2328 non-null   int64 
 1   index          2328 non-null   int64 
 2   talk_time      2328 non-null   object
 3   battery_power  2328 non-null   object
 4   pc             2328 non-null   object
 5   three_g        2328 non-null   object
 6   mobile_wt      2328 non-null   object
 7   px_width       2328 non-null   object
 8   sc_h           2328 non-null   object
 9   sc_w           2328 non-null   object
 10  m_dep          2328 non-null   object
 11  touch_screen   2328 non-null   object
 12  fc             2328 non-null   object
 13  four_g         2328 non-null   object
 14  hgv            2328 non-null   object
 15  price_range    2328 non-null   object
 16  blue           2328 non-null   object
 17  n_cores        2328 non-null   object
 18  wifi           2328 non-nul

In [37]:
print( 'Duplicated Registers: ' + str( ds["index"].duplicated().sum() ))
print( 'Unique Registers: ' + str( len ( ds["index"].unique() ) ) )
print(f"Total missing values in the dataset: {ds.isnull().sum().sum()}")

Duplicated Registers: 0
Unique Registers: 2328
Total missing values in the dataset: 0


### Elimination of features that do not provide information

The columns _Unnamed:0_ and _index_ are eliminated since they are only numerical identifiers for the cell phone models consulted by Pedro Perez.

The _werf_ column is eliminated since it only has a single value in all the records of the dataset.

In [38]:
ds.drop(['Unnamed: 0', 'index'], axis=1, inplace=True)


## 4. Save intermediated transformed data

Data without:
- Duplicates
- Missings values
- Non informative data for our taks

In [43]:
ds.to_csv('../data/interim/CellPhoneDs_clean.csv', index = False)


## 5. Partial Results

### Initial raw data:
74790 rows, 25 Columns , memory usage: 10.9 MB

### After cleaning process
2328 rows , 22 Columns , memory usage: 253.8 kB

***
